p2p/src/main/kotlin/maru/p2p/MaruPeerManager.kt

```diff
@@ -108,29 +121,70 @@ class MaruPeerManager(
   }
 
   override fun onConnect(peer: Peer) {
-    // TODO: here we could check if we want to be connected to that peer
-    if (p2pNetwork.peerCount > maxPeers) {
+    if (connectedPeers.size >= maxPeers) {
       // TODO: We could disconnect another peer here, based on some criteria
       peer.disconnectCleanly(DisconnectReason.TOO_MANY_PEERS)
+      return
     }
     val maruPeer = maruPeerFactory.createMaruPeer(peer)
-    connectedPeers.put(peer.id, maruPeer)
-    if (maruPeer.connectionInitiatedLocally()) {
-      maruPeer.sendStatus()
-    } else {
-      maruPeer.scheduleDisconnectIfStatusNotReceived(p2pConfig.statusUpdate.timeout)
-    }
+    statusExchangingMaruPeers[peer.id] = maruPeer
+    maruPeer
+      .awaitInitialStatus()
+      .thenApply { status ->
+        if (!status.forkIdHash.contentEquals(forkidHashProvider.currentForkIdHash())) {
+          log.debug(
+            "Peer={} has a different forkIdHash={} than expected={}. Disconnecting.",
+            peer.id,
+            status.forkIdHash,
+            forkidHashProvider.currentForkIdHash(),
+          )
+          maruPeer.disconnectCleanly(DisconnectReason.IRRELEVANT_NETWORK)
+        } else if (syncStatusProvider.getCLSyncStatus() == CLSyncStatus.SYNCING &&
+          status.latestBlockNumber < syncStatusProvider.getSyncTarget()!! - blocksBehindThreshold
+        ) {
+          log.debug(
+            "Peer={} is too far behind our target block number={} (peer's block number={}). Disconnecting.",
+            peer.id,
+            syncStatusProvider.getSyncTarget(),
+            status.latestBlockNumber,
+          )
+          maruPeer.disconnectCleanly(DisconnectReason.TOO_MANY_PEERS) // there is no better reason available
+        } else if (syncStatusProvider.getCLSyncStatus() == CLSyncStatus.SYNCED &&
+          status.latestBlockNumber < beaconChain.getLatestBeaconState().latestBeaconBlockHeader.number &&
```

Comment:
Can this condition be moved into a function so it's clearer. It's also used in tooManyPeersBehind I think

Reply:
That would probably look better :-)